packages/utils/src/dal/mikro-orm/utils.ts

```diff
@@ -143,9 +158,341 @@ export const mikroOrmUpdateDeletedAtRecursively = async <
   }
 }
 
+// SERIALIZATION UTILS //
+
+function isVisible<T extends object>(
+  meta: EntityMetadata<T>,
+  propName: string,
+  options: SerializeOptions<T, any>
+): boolean {
+  if (options.populate === true) {
+    return options.populate
+  }
+
+  if (
+    Array.isArray(options.populate) &&
+    options.populate?.find(
+      (item) => item === propName || item.startsWith(propName + ".")
+    )
+  ) {
+    return true
+  }
+
+  if (options.exclude?.find((item) => item === propName)) {
+    return false
+  }
+
+  const prop = meta.properties[propName]
+  const visible = prop && !prop.hidden
+  const prefixed = prop && !prop.primary && propName.startsWith("_") // ignore prefixed properties, if it's not a PK
+
+  return visible && !prefixed
+}
+
+function isPopulated<T extends object>(
+  entity: T,
+  propName: string,
+  options: SerializeOptions<T, any>
+): boolean {
+  if (
+    typeof options.populate !== "boolean" &&
+    options.populate?.find(
+      (item) => item === propName || item.startsWith(propName + ".")
+    )
+  ) {
+    return true
+  }
+
+  if (typeof options.populate === "boolean") {
+    return options.populate
+  }
+
+  return false
+}
+
+export class EntitySerializer {
+  static serialize<T extends object, P extends string = never>(
+    entity: T,
+    options: SerializeOptions<T, P> & { preventCircularRef?: boolean } = {},
+    parent?: object
+  ): EntityDTO<Loaded<T, P>> {
+    const wrapped = helper(entity)
+    const meta = wrapped.__meta
+    let contextCreated = false
+
+    if (!wrapped.__serializationContext.root) {
+      const root = new SerializationContext<T>()
+      SerializationContext.propagate(
+        root,
+        entity,
+        (meta, prop) =>
+          meta.properties[prop]?.reference !== ReferenceType.SCALAR
+      )
+      contextCreated = true
+    }
+
+    const root = wrapped.__serializationContext.root!
+    const ret = {} as EntityDTO<Loaded<T, P>>
+    const keys = new Set<string>(meta.primaryKeys)
+    Object.keys(entity).forEach((prop) => keys.add(prop))
+    const visited = root.visited.has(entity)
+
+    if (!visited) {
+      root.visited.add(entity)
+    }
+
+    ;[...keys]
```

Comment:
Just a NIT: We can maybe break down this loop and the one below using helper functions, so it's easier to understand the flow of data

Reply:
The idea was to keep it as close as possible to what mikro orm has, the idea behind to not divert too much and be able in to read both side by side to see the diff at any time, i would prefer not break it down to smaller function for that reason in that particular case. I can add some tests though

Reply:
@adrien2p ah so most of the code is taken from mikroORM? Let's maybe add a comment in case someone else needs to modify it? I am ok with that then 👍

Reply:
Yes exactly, i ve taken it and added the necessary logic for our needs

Reply:
Would it make sense to encapsulate our custom logic to make it very explicit? 

Either with comments, or in a separate method - you know best what would be fitting.

Reply:
Yes i ll add comments and tests